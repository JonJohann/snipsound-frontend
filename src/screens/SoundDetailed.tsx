import React, { useEffect, useState } from 'react';
import { Button, View, Text, StyleSheet, TouchableOpacity} from 'react-native';
import { Sound } from '../components/misc/SingleSound'
import { Audio, AVPlaybackStatus } from 'expo-av';
import { withSafeAreaInsets } from 'react-native-safe-area-context';
import { styles } from '../styles/styles'
import { FontAwesome } from '@expo/vector-icons';
import { Entypo } from '@expo/vector-icons';
import { dateAndTime } from '../components/misc/SingleSound'
import Divider from '../components/misc/Divider';

export interface Function {
    callback: () => void
}

// Detailed view of each Snipsound
export default function SoundDetailed(props: Sound & Function) {

    const [playing, setPlaying] = useState(false)
    const initialSound = new Audio.Sound();
    const [sound] = useState(initialSound)
    const date = new Date(props.date)

    const stopPlaying = () => {
        if (playing) {
            sound.stopAsync()
            sound.unloadAsync()
        }
    }

    const playOrPause = async () => {
        try {
            if (!playing) {
                await sound.loadAsync({ uri: 'http://84.202.202.48:3074/' + props.filename });
                await sound.setIsLoopingAsync(true)
                await sound.playAsync();
                setPlaying(true)
            }
            else {
                await sound.stopAsync();
                await sound.unloadAsync();
                setPlaying(false)
            }
        } catch (error) {
            console.log(error)
        }
    }

    return (
        <View style={styles.centeredView}>
            <View style={styles.detailedView}>
                <Text style={[styles.title, {marginBottom: 10}]}>{props.title}</Text>
                <Text style={styles.detailedText}>By {props.owner}</Text>
                <Text style={styles.detailedText}> Category: {props.category} </Text>
                <Text style={styles.detailedText}> Posted on: {dateAndTime(date)} </Text>
                <Divider/>
                <Text style={styles.description}>{props.description}</Text>
                <Divider/>
                <View style={styles.buttonRow}>
                    <TouchableOpacity onPress={playOrPause} style={styles.symbols}>
                        <FontAwesome style={styles.symbols} name={playing ? "pause" : "play"} size={50} />
                    </TouchableOpacity>
                    <TouchableOpacity onPress={() => { stopPlaying(); props.callback() }} style={styles.symbols}>
                        <Entypo name="cross" size={70} color="black" />
                    </TouchableOpacity>
                </View>
            </View>
        </View>
    )
}
