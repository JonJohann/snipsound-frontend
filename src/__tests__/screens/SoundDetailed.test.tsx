import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import SoundDetailed from '../../screens/SoundDetailed';
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

describe('<SoundDetailed />', () => {
    const initialState = {
        sounds: {
                posts: [],
                totalPages: 1,
                status: 'idle',
                error: "",
                user: ""
        },
        params: {
            sortby: "-date",
            search: "",
            category: "",
            limit: 5
        }
    }
    const callback = jest.fn()
    const initialPost = {
        _id: "",
        date: "",
        title: "",
        owner: "",
        filename: "",
        category: "",
        description: "",
        likes: 0,
    }

    beforeEach(() => {
        jest.useFakeTimers()
    })
    const mockStore = configureStore()
    let store = mockStore(initialState)
    it('has 2 children', () => {
        const tree = renderer.create(<Provider store={store}><SoundDetailed callback={callback} {...initialPost}/></Provider>).toJSON() as ReactTestRendererJSON;
        expect(tree!.children!.length).toBe(1);
    });

    it('renders correctly', () => {
        const tree = renderer.create(<Provider store={store}><SoundDetailed callback={callback} {...initialPost}/></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});