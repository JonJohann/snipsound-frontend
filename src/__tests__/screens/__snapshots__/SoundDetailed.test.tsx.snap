// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SoundDetailed /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "marginTop": 20,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255,255,255, 0.96)",
        "borderRadius": 20,
        "padding": 35,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Manrope_400Regular",
            "fontSize": 30,
            "fontWeight": "bold",
          },
          Object {
            "marginBottom": 10,
          },
        ]
      }
    >
      
    </Text>
    <Text
      style={
        Object {
          "fontFamily": "Manrope_400Regular",
          "fontSize": 20,
          "marginVertical": 2,
          "textAlign": "center",
        }
      }
    >
      By 
      
    </Text>
    <Text
      style={
        Object {
          "fontFamily": "Manrope_400Regular",
          "fontSize": 20,
          "marginVertical": 2,
          "textAlign": "center",
        }
      }
    >
       Category: 
      
       
    </Text>
    <Text
      style={
        Object {
          "fontFamily": "Manrope_400Regular",
          "fontSize": 20,
          "marginVertical": 2,
          "textAlign": "center",
        }
      }
    >
       Posted on: 
      NaN.NaN.NaN at NaN:NaN
       
    </Text>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "borderTopColor": "#F194FF",
          "borderTopWidth": 1,
          "marginVertical": 20,
          "width": 300,
        }
      }
    />
    <Text
      style={
        Object {
          "borderColor": "#acacac",
          "fontFamily": "Manrope_400Regular",
          "fontSize": 20,
          "paddingHorizontal": 10,
          "textAlign": "center",
        }
      }
    >
      
    </Text>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "borderTopColor": "#F194FF",
          "borderTopWidth": 1,
          "marginVertical": 20,
          "width": 300,
        }
      }
    />
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "margin": 10,
            "opacity": 1,
          }
        }
      >
        <Text />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "margin": 10,
            "opacity": 1,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
